Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx

```diff
@@ -0,0 +1,239 @@
+    // be catched in the main function, hence, this "if" is no longer necessary
+    LOGP(warning, "Sector size amount of data points) in current run is 0!");
+    LOGP(warning, "mMapOfAllDigits with 0's is generated in order to prevent a segementation fault.");
+
+    return;
+  }
+
+  for (const auto& cli : eventSector) {
+    Times.emplace_back(cli.getTimeStamp());
+    Pads.emplace_back(cli.getPad());
+    Rows.emplace_back(cli.getRow());
+  }
+  // here a "mMapOfAllDigits" large enough to contain all data points of the envents is
+  // initialized and then filled
+  mMapOfAllDigits = std::array<std::array<std::array<float, MaxPads>, MaxRows>, MaxTimes>{};
+  for (const auto& digit: eventSector) {
+    mMapOfAllDigits[digit.getTimeStamp()][digit.getRow()][digit.getPad()] = digit.getChargeFloat();
+  }
+}
+
+//#################################################
+
+// Function to update the temporal cluster
+void KrBoxClusterFinder::updateTempClusterFinal()
+{
+  const float oneOverQtot = 1. / mTempCluster.totCharge;
+  mTempCluster.meanPad *= oneOverQtot;
+  mTempCluster.sigmaPad *= oneOverQtot;
+  mTempCluster.meanRow *= oneOverQtot;
+  mTempCluster.sigmaRow *= oneOverQtot;
+  mTempCluster.meanTime *= oneOverQtot;
+  mTempCluster.sigmaTime *= oneOverQtot;
+  mTempCluster.sigmaPad = std::sqrt(std::abs(mTempCluster.sigmaPad - mTempCluster.meanPad * mTempCluster.meanPad));
+  mTempCluster.sigmaRow = std::sqrt(std::abs(mTempCluster.sigmaRow - mTempCluster.meanRow * mTempCluster.meanRow));
+  mTempCluster.sigmaTime = std::sqrt(std::abs(mTempCluster.sigmaTime - mTempCluster.meanTime * mTempCluster.meanTime));
+}
+
+// Function to update the temporal cluster.
+void KrBoxClusterFinder::updateTempCluster(float tempCharge, int tempPad, int tempRow, int tempTime)
+{
+  if (tempCharge >= mQThreshold) {
+    mTempCluster.size += 1;
+    mTempCluster.totCharge += tempCharge;
+
+    mTempCluster.meanPad += tempPad * tempCharge;
+    mTempCluster.sigmaPad += tempPad * tempPad * tempCharge;
+
+    mTempCluster.meanRow += tempRow * tempCharge;
+    mTempCluster.sigmaRow += tempRow * tempRow * tempCharge;
+
+    mTempCluster.meanTime += tempTime * tempCharge;
+    mTempCluster.sigmaTime += tempTime * tempTime * tempCharge;
+    if (tempCharge > mTempCluster.maxCharge) {
+      mTempCluster.maxCharge = tempCharge;
+    }
+  } else {
+    LOGP(warning, "Update cluster was called but current charge is below mQThreshold");
+  }
+}
+
+// This function finds and evaluates all clusters in a 3D mMapOfAllDigits generated by the
+// mMapOfAllDigitsCreator function, this function also updates the cluster tree
+std::vector<std::tuple<int, int, int>> KrBoxClusterFinder::findLocalMaxima()
+{
+  std::vector<std::tuple<int, int, int>> localMaximaCoords;
+  // loop over whole mMapOfAllDigits the find clusers
+  for (size_t iRow = 0; iRow < MaxRows; iRow++) {
+    for (size_t iPad = 0; iPad < MaxPads; iPad++) {
+      for (size_t iTime = 0; iTime < MaxTimes; iTime++) {

```

Comment:
Is there a specific reason for order in the nesting? I think it would make more sense to have the inner most loop over the inner most array elemets to avoid jumping in memory. If time should be the inner most loop then it might be better to also change it in the definition of `mMapOfAllDigits`

```suggestion
  for (size_t iTime = 0; iTime < MaxTimes; iTime++) {
    for (size_t iRow = 0; iRow < MaxRows; iRow++) {
      for (size_t iPad = 0; iPad < MaxPads; iPad++) {

```

Reply:
No specific reason. Done.